### ISSUE TITLE
Fix memory limitation

### ISSUE THREAD

**Issue by ehem** [dornerworks/xen-rpi4-builder#3]
Limiting system memory to 1GB as documented is distinctly sub-optimal and needs fixing.

My understanding of 0001-Add-RPi4-bcm2711-device-tree-for-Xen-64-bit-Linux.patch is you're also forcing Domain 0 to only get 512MB.  This is reasonable under many, but not all circumstances.  A device where X is being run in Domain 0 may want more there.

Thanks for documenting the limitation though as that is important to know.

**Comment by tklengyel** [comment]
Can we get some more information about where the 1GB memory limitation comes from? I see that in config.txt rpixen.sh sets `total_mem=1024`. My rpi4 has 4gb memory but when I change this to 4096 dom0 doesn't boot. With changing it to something a little bit higher like `1096` it boots but Xen still says available total_memory is `1020`. I can't really make heads or tails of this behavior.

**Comment by stewdk** [comment]
I don't know exactly what the issue is, but I suspect that it is related to the fact that certain DMA-capable peripherals can only address the first GB (https://patchwork.kernel.org/cover/11141719/).

As you've also done, I’ve tried `total_mem=4096` - Xen successfully passes control to dom0, but Linux does not recognize the SD card in this case. The Raspberry Pi 4 is the first model of Raspberry Pi to support 4GB, but they've reused many of the peripherals and drivers from previous generations where there was only 1GB. Is it possible some assumptions were made during development of various drivers about the addressing space being limited to 1GB? More investigation is needed.

I've been intending to try out the upstream kernel now that some initial RPi4 support has been merged in Linux 5.5.

**Comment by ianmclinden** [comment]
I'm able to allocate more memory to dom0 by the following:

```
arch/arm64/boot/dts/broadcom/bcm2711-rpi-4-b-xen.dts`
-- xen,xen-bootargs = "console=dtuart dtuart=/soc/serial@7e215040 sync_console dom0_mem=512M bootscrub=0"
++ xen,xen-bootargs = "console=dtuart dtuart=/soc/serial@7e215040 sync_console dom0_mem=3072m bootscrub=0"
---
bootfiles/config.txt
-- total_mem=1024
++ total_mem=4096
```

While this increases the memory available to dom0, it doesn't increase the effective guest memory, as ballooning doesn't work out of the box.

### Boot Logs
- [Stock boot UART log](https://github.com/dornerworks/xen-rpi4-builder/files/4411653/xen-rpi4-stock.log)
- [3G boot UART log](https://github.com/dornerworks/xen-rpi4-builder/files/4411652/xen-rpi4-3g.log)

### Update (04/09/2020)
Booting with these changes seems to cause USB bus limitations. Specifically, no usb-bus devices are showing up.

**Comment by stewdk** [comment]
See https://lists.xenproject.org/archives/html/xen-devel/2020-05/msg00960.html

**Comment by ehem** [comment]
> See https://lists.xenproject.org/archives/html/xen-devel/2020-05/msg00960.html

That is now in Xen's main branch as commit 437b0aa06a014ce174e24c0d3530b3e9ab19b18b.  The commits are also compatible with 4.13.

**Comment by stewdk** [comment]
This is resolved with Xen 4.15 and Linux 5.10.
